gateway-service/src/main/java/com/ca/mfaas/gateway/filters/post/PageRedirectionFilter.java

```diff
@@ -0,0 +1,205 @@
+ * <li>The url can be matched to gateway url</li>
+ * </ul>
+ */
+@Slf4j
+public class PageRedirectionFilter extends ZuulFilter implements RoutedServicesUser {
+
+    private final DiscoveryClient discovery;
+    private final Map<String, RoutedServices> routedServicesMap = new HashMap<>();
+    private final ConcurrentHashMap<String, String> routeTable = new ConcurrentHashMap<>();
+    private final TransformService transformService;
+
+    /**
+     * Constructor
+     *
+     * @param discovery discovery client
+     * @param hostname  hostname of gateway
+     * @param port      port number of gateway
+     * @param scheme    scheme of gateway
+     */
+    public PageRedirectionFilter(DiscoveryClient discovery, String hostname, String port, String scheme) {
+        this.discovery = discovery;
+        GatewayConfigProperties gatewayConfigProperties = GatewayConfigProperties.builder()
+            .scheme(scheme)
+            .hostname(hostname + ":" + port)
+            .build();
+        transformService = new TransformService(gatewayConfigProperties);
+    }
+
+    /**
+     * @return true if status code is 3XX
+     */
+    @Override
+    public boolean shouldFilter() {
+        RequestContext context = RequestContext.getCurrentContext();
+        int status = context.getResponseStatusCode();
+        return HttpStatus.valueOf(status).is3xxRedirection();
+    }
+
+    @Override
+    public String filterType() {
+        return POST_TYPE;
+    }
+
+    @Override
+    public int filterOrder() {
+        return SEND_RESPONSE_FILTER_ORDER - 1;
+    }
+
+    /**
+     * When the filter runs, it first finds the Location url in cache. If matched url can be found in cache, it then replaces Location with the matched url.
+     * If not, the filter will find the matched url in Discovery Service. If matched url can be found in Discovery Service, the filter will put the matched
+     * url to cache, and replace Location with the matched url
+     *
+     * @return null
+     */
+    @Override
+    public Object run() {
+        RequestContext context = RequestContext.getCurrentContext();
+        Optional<Pair<String, String>> locationHeader = context.getZuulResponseHeaders()
+            .stream()
+            .filter(stringPair -> LOCATION.equals(stringPair.first()))
+            .findFirst();
+
+        if (locationHeader.isPresent()) {
+            String location = locationHeader.get().second();
+            //find matched url in cache
+            String transformedUrl = foundUrlInTable(location);
+            if (transformedUrl != null) {
+                transformLocation(locationHeader.get(), transformedUrl);
+            } else {
+                //find matched url in Discovery Service
+                Optional transformedUrlOp = getMatchedUrlFromDS(location);
+                if (transformedUrlOp.isPresent()) {
+                    transformedUrl = (String) transformedUrlOp.get();
```

Comment:
```suggestion
                    transformedUrl = transformedUrlOp.get();
```

Reply:
Hi Ilkin,
Thanks for comments, I've modified the code in latest commit